src/lib/Microsoft.Health.Common/Auth/Provider/IAzureExternalIdentityCredentialProvider.cs

```diff
@@ -0,0 +1,12 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+namespace Microsoft.Health.Common.Auth
+{
+    // needed to differentiate cases where there are more than one IAzureCredentialProvider
+    public interface IAzureExternalIdentityCredentialProvider : IAzureCredentialProvider
```

Comment:
I was thinking that IAzureExternalIdentityCredentialProvider would need to be in OSS so that our event hub and our fhir service clients can be configured to toggle between managed identity token providers. In OSS we will inject only an IAzureCredentialProvider since all the clients will use the same identity, but for PaaS I was thinking we could inject both an IAzureCredentialProvider for the internal resources and an IAzureExternalIdentityCredentialProvider for external resources and the client would use a hierarchy like this to get the token provider (currently in [NormalizationExtensions.cs](https://github.com/microsoft/iomt-fhir/pull/251/files#diff-f43a75fdb17e394368158fcb0245b03abecec92a4e76044e28ab0c94fe310b12) and FhirTransformationExtensions.cs):

```
  // Get IAzureExternalIdentityCredentialProvider if it exists, else use IAzureCredentialProvider
  var externalMiTokenProvider = sp.GetService<IAzureExternalIdentityCredentialProvider>();
  var serviceMiTokenProvider = sp.GetService<IAzureCredentialProvider>();
  var tokenProvider = externalMiTokenProvider ?? serviceMiTokenProvider;
```

Reply:
Sorry, I wrote the comment after trying to find references (for some reason find reference didn't work on my local copy) but I did see this later and I forgot to update this comment.

This works I am just trying to think if there is a cleaner way to accomplish this.  My main issue is we are assuming the IAzureCredentialProvider is an appropriate swap for the IAzureExternalIdentityProvider when the latter is not defined.  Some other issues are the IAzureExternalIdentityCredentialProvider and IAzureCredentialProvider are identical but we add the new type to support multiple instances in DI.  It also is limited in how it scales, this works when we only need to support two concurrent identity providers, if we ever need to support a third or more it starts to break down.  

One rough idea I had is we could introduce something like this to the Microsoft.Health.Common.Auth library.

The interface and class would look like this:
```
    public interface IAzureCredentialProvider<in T>
    {
        Azure.Core.TokenCredential GetCredential();
    }
```

```
    public class AzureCrendentialProvider<T> : IAzureCredentialProvider<T>
    {
        private readonly IAzureCredentialProvider _credentialProvider;

        public AzureCrendentialProvider(IAzureCredentialProvider azureCredentialProvider)
        {
            _credentialProvider = azureCredentialProvider;
        }

        public TokenCredential GetCredential()
        {
            return _credentialProvider.GetCredential();
        }
    }
```

The classes then would be updated to use IAzureCredentialProvider<T> where the T would be their own type and allow you to define credential providers per type.

We could have extension methods that look like this:
```
        public static IServiceCollection AddDefaultNormalizedEventReaderCredentialProvider(this IServiceCollection services)
        {
            services.AddSingleton<IAzureCredentialProvider<IEventProcessorClientFactory>>(sp => new AzureCredentialProvider<IEventProcessorClientFactory>(sp.GetRequiredService<AzureCredentialProvider>()));
            return services;
        }
```
This would be invoked in OSS where as we could have something like this for PaaS.

```
        public static IServiceCollection AddExternalNormalizedEventReaderCredentialProvider(this IServiceCollection services)
        {
            services.AddSingleton<IAzureCredentialProvider<IEventProcessorClientFactory>>(sp => new AzureCredentialProvider<IEventProcessorClientFactory>(sp.GetRequiredService<ExternalIdentityCredentialProvider>()));
            return services;
        }
```

We would extend this for the event reader and FHIR client as well . Thoughts?

Reply:
I am on the fence about switching to IAzureCredentialProvider<in T>. It is definitely cleaner in some respects for the event reader and fhir client since those would not need to be modified when adding >=2 credential providers. My main concern is with the extension methods - we would need to create specific methods and classes in OSS and PaaS and it would not be consolidated. 

I did also try implementing it anyway and found that a bit more refactoring was needed than expected and am leaning towards not implementing (at least not in this PR) due to number of additional changes